keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1301 @@
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `endofterm: true` on the leader `Lease`, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: some-custom-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
+
+In an HA configuration, the Coordinated Leader Election Controller will have its
+own lease similar to how other leader elected controllers behaves today. It will
+be responsible for renewing its own lease and gracefully shutdown if the lease
+is expired. Only one instance of the coordinated leader election controller will
+be active at a time, and this prevents instances of the coordinated leader
+election controller from interfering with each other. Unlike in KCM, the
+coordinated leader election controller must gracefully shutdown and restart as
+it will be running in the kube-apiserver and calling `os.Exit()` is not an
+option.
+
+### Coordinated Lease Lock
+
+A new `resourceLock` type of `coordinatedleases`, and `CoordinatedLeaseLock`
+implementation of `resourcelock.Interface` will be added to client-go that:
+
+- Creates Identity Lease when ready to be Leader
+- Renews identity lease periodically
+- Watches its identity lease for the `coordination.k8s.io/pending-ack` annotation and updates to remove it
+- Watches Leader Lease, waiting to be elected leader by the Coordinated Election
+  Controller
+- When it becomes leader:
+  - Perform role of active component instance
+  - Renew leader lease periodically
+  - Stop renewing if lease is marked `coordination.k8s.io/end-of-term: true`
+- If leader lease expires:
+  - Shutdown (yielding leadership) and restart as a candidate component instance
+
+```mermaid
+flowchart TD
+    A[Started] -->|Create Identity Lease| B
+    B[Candidate] --> |Elected| C[Leader]
+    C --> |Renew Leader Lease| C
+    C -->|End of Term / Leader Lease Expired| D[Shutdown]
+    D[Shutdown] -.-> |Restart| A
+```
+
+### Renewal Interval and Performance
+The leader lease will have similar renewal interval and duration (2s and 15s).
+
+For component leases, keeping a short renewal interval will add many unnecessary writes to the apiserver.
+The component leases renewal interval will default to 30s.
```

Comment:
Given that we will be setting the `'ending-ack` annotation and waiting for explicit ACK on it, the 30s reneval period seems very low to me. I would rather target something in the order of 30m or sth - given the ACk mechanism below, we don't gain pretty much anything from those frequent periodic refreshes.

Reply:
I've set it to 5 minutes for now because that's the default resync period for the informer cache. Is that fine?